iothub/device/src/Pipeline/ConnectionStateDelegatingHandler.cs

```diff
@@ -0,0 +1,382 @@
+
+                                        // Create a new cancelaltion token source that will be used for reconnection recovery attempts.
+                                        _handleDisconnectCts = new CancellationTokenSource();
+
+                                        // Send the request for transport close notification.
+                                        _transportClosedTask = HandleDisconnectAsync();
+                                    }
+                                    catch (Exception ex) when (!Fx.IsFatal(ex))
+                                    {
+                                        if (Logging.IsEnabled)
+                                            Logging.Error(this, ex, nameof(HandleDisconnectAsync));
+
+                                        HandleConnectionStatusExceptions(ex, true);
+                                        _clientTransportStateMachine.MoveNext(ClientStateAction.OpenFailure);
+                                        //SetClientTransportStatus(ClientTransportState.Closed);
+
+                                        throw;
+                                    }
+                                }
+                            }
+                            finally
+                            {
+                                _clientOpenSemaphore?.Release();
+                            }
+
+                            break;
+                        }
+                }
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, cancellationToken, nameof(OpenAsync));
+            }
+        }
+
+        public override async Task SendTelemetryAsync(TelemetryMessage message, CancellationToken cancellationToken)
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, message, cancellationToken, nameof(SendTelemetryAsync));
+
+            try
+            {
+                using var operationCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _cancelPendingOperationsCts.Token);
+
+                switch (_clientTransportStateMachine.GetCurrentState())
+                {
+                    case ClientTransportState.Opening:
+                        await _clientOpenSemaphore.WaitAsync(operationCts.Token).ConfigureAwait(false);
+
+                        try
+                        {
+                            if (_clientTransportStateMachine.GetCurrentState() != ClientTransportState.Open)
+                            {
+                                throw new InvalidOperationException($"The client connection must be opened before operations can begin. Call '{nameof(OpenAsync)}' and try again.");
+                            }
+                        }
+                        finally
+                        {
+                            _clientOpenSemaphore?.Release();
+                        }
```

Comment:
I suspect this would be a reusable snippet for other override methods too, so prime candidate for moving to a private method?

Reply:
yes, that's correct